autoagora_agents/algorithm.py

```diff
@@ -0,0 +1,432 @@
+# Copyright 2022-, Semiotic AI, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+from abc import ABC, abstractmethod
+
+import numpy as np
+import torch
+from torch import optim
+
+import experiment
+from autoagora_agents import buffer
+from autoagora_agents.distribution import distributionfactory
+
+
+class Algorithm(ABC):
+    def __init__(self, *, group: str, i: int) -> None:
+        self.niterations = 0
+        self.nupdates = 0
+        self.group = group
+        self.i = i
+        self.name = f"{group}_{i}"
+
+    def reset(self) -> None:
+        self.niterations = 0
+
+    def update(self):
+        self.nupdates += 1
+
+    @abstractmethod
+    def __call__(
+        self,
+        *,
+        observation: np.ndarray,
+        action: np.ndarray,
+        reward: float,
+        done: bool,
+    ) -> np.ndarray:
+        pass
+
+    @staticmethod
+    def advantage(rewards: torch.Tensor) -> torch.Tensor:
+        """Compute a simple advantage estimate.
+
+        In effect, this is just standardising the samples to N(0, 1)
+
+        Arguments:
+            rewards (torch.Tensor): The reward-history using which to compute the advantage
+
+        Returns:
+            torch.Tensor: The advantage estimate
+        """
+        std = rewards.std()
+        if torch.isnan(std) or std == 0:
+            adv = rewards
+        else:
+            adv = (rewards - rewards.mean()) / rewards.std()
+        return torch.unsqueeze(adv, dim=1)
+
+
+class PredeterminedAlgorithm(Algorithm):
+    """Change to a particular value at a given timestamp.
+
+    Attributes:
+        timestamps (list[int]): The timestamps at which to change the outputted value.
+            Must start with 0.
+        vals (list[np.ndarray]): The values outputted.
+    """
+
+    def __init__(
+        self, *, group: str, i: int, timestamps: list[int], vals: list[np.ndarray]
+    ) -> None:
+        super().__init__(group=group, i=i)
+        if timestamps[0] != 0:
+            raise ValueError("The first timestamp must be 0.")
```

Comment:
I would also check if lengths of both lists are the same

Reply:
7ebc3283b9df6d4455d0ee807ae9ab1de7b18a59